### ISSUE TITLE
Paire (sym) issue in the transport equation

### ISSUE THREAD

**Issue by semi-h** [xcompact3d/x3d2#14]
This issue is about the infamous 'paire' in Incompact3D (renamed as sym in the new framework). And then it links to our discussion on Tuesday with @pbartholomew08 and @Nanoseb.

In transport equation we call the directional derivatives paying attention to this 'paire' and pass the right preprocessed coefficient arrays (`ff, fs, fw` and `sf, ss, sw`).

We have found a solution to this a while ago with @slaizet. Here is the plan:
Assume we have a function that obtains all the derivatives for the transport_equation for all three momentum equations in one direction and writes them back to 3 solution fields;

`call transeq_dir(du, dv, dw, u, v, w, ...)`

in x dir it has to be (see https://github.com/xcompact3d/Incompact3d/blob/master/src/transeq.f90#L117-L122)
```
du = der_paired(u*u) + u*der_nopair(u)
dv = der_nopair(u*v) + u*der_paired(v)
dw = der_nopair(u*w) + u*der_paired(w)
```
And in y dir paired/non paired stuff looks a bit different https://github.com/xcompact3d/Incompact3d/blob/master/src/transeq.f90#L190-L195
```
du = der_nopair(v*u) + v*der_paired(u)
dv = der_paired(v*v) + v*der_nopair(v)
dw = der_nopair(v*w) + v*der_paired(w)
```
So we know how the final calls should look like. To simplify this process, lets call the function that executes x directional derivatives by rearranging the argument order 

 `call transeq_dir(dv, du, dw, v, u, w, ...)`

it'll do the following for us (switch all the u's by v's and all the v's by u's)
```
dv = der_paired(v*v) + v*der_nopair(v)
du = der_nopair(v*u) + v*der_paired(u)
dw = der_nopair(v*w) + v*der_paired(w)
```
(lets reorder equations to make them look more natural);
```
du = der_nopair(v*u) + v*der_paired(u)
dv = der_paired(v*v) + v*der_nopair(v)
dw = der_nopair(v*w) + v*der_paired(w)
```
So we're just changing the order of fields we pass and then the paired/non paired are always in one particular order inside the function. This way we hide paire specific stuff in the transport equation. 
**Does anyone foresee any isues resulting from this approach?**

Back to the discussion on Tuesday, the main thing was where we switch the order or `u, v, w`. The current state in [Backends PR ](https://github.com/xcompact3d/x3d2/pull/7) is that we call transeq_dir in all directions in `u, v, w` order as shown below, and then in CUDA or OpenMP backends we carry out the shift and call the transeq_dir as `v, u, w`in y and `w, u, v` in z.
```
   subroutine transeq(self, du, dv, dw, u, v, w)
      implicit none

      class(base_backend_t) :: self
      class(field_t), intent(out) :: du, dv, dw
      class(field_t), intent(in) :: u, v, w

      ! x derivatives in the transport equation
      call self%transeq_x(du, dv, dw, u, v, w, self%xdirps)

      ! get u_y, v_y, w_y by rearranging u, v, w
      ! similar to the x direction, obtain derivatives in y.
      call self%transeq_y(du_y, dv_y, dw_y, u_y, v_y, w_y, self%ydirps)

      ! get u_z, v_z, w_z by rearranging u, v, w
      ! get the derivatives in z
      call self%transeq_z(du_z, dv_z, dw_z, u_z, v_z, w_z, self%zdirps)
```
This is actually what we agreed on Tuesday but thinking on this a bit more I'm not sure if its the right way. If we carry out this argument order rearrangement in the CUDA/OpenMP backends, we'll be doing something common at a lower level. I think this is conceptually wrong and we should do the reordering at base backend level and call the transeq_dir function as below, what do you think?

```
      call self%transeq_y(dv_y, du_y, dw_y, v_u, u_y, w_y, self%ydirps)
      call self%transeq_z(dw_z, du_z, dv_z, w_z, u_z, v_z, self%zdirps)
```

**Comment by slaizet** [comment]
No issue with the proposed approach!

**Comment by mathrack** [comment]
Yes, the wall-normal velocity and the wall-parallel velocities are treated differently.
